### DISCUSSION TITLE
Adsense in Next.js is possible?

### DISCUSSION THREAD

**Discussion by pedrobonifacio** [vercel/next.js#14395] [Help]
How to implement google ads(adsense) in a Next.js application?

**Comment by SarKurd** [comment]
Add this to the `head` in `_document.js`

```html
<script
            async
            src="https://pagead2.googlesyndication.com/pagead/js/adsbygoogle.js"
          ></script>
```

and use it like this (in my case i created a re-usable ad component)

```js
  
import React, { useEffect } from "react";

const AdBanner = () => {
  useEffect(() => {
    try {
      (window.adsbygoogle = window.adsbygoogle || []).push({});
    } catch (err) {
      console.log(err);
    }
  }, []);

  return (
    <ins
      className="adsbygoogle adbanner-customize"
      style={{
        display: "block"
      }}
      data-ad-client=<your-client-id>
      data-ad-slot=<slot-id>
    />
  );
};

export default AdBanner;
```

**Reply by EfiGordon** [reply]
I dont have _document.js, I have index.js or  _app.js?

**Reply by SarKurd** [reply]
> I dont have _document.js, I have index.js or _app.js?

create one, https://nextjs.org/docs/advanced-features/custom-document

**Reply by kas** [reply]
~~Do you know why adsbygoogle sometimes doesn't get initialized before `(adsbygoogle = window.adsbygoogle || []).push({})` runs? I know the try catch will catch it, but I'm curious why it fails sometimes in the first place.~~

~~I can repro it by pressing control + shift + r to reload the page.~~

~~I can also repro it by visiting my website from a new device that hasn't cached it yet (e.g., my phone). I think that means the first time the website is loaded, the ad won't display.~~

It turns out you must use `(window.adsbygoogle = window.adsbygoogle || []).push({})`, the extra "window" in the beginning makes the difference. Thanks!!

**Reply by alex-r89** [reply]
hi @SarKurd thanks for this.

How do you implement this component? Do you just drop your `AdBanner` into places and let google work out the heights and widths? Or do you wrap this in a fixed height/width container?

I get an error using this code, `No slot size for availableWidth=0"` but I have this component inside a container with a fixed height and width so unsure why its not filling the space responsively. 

EDIT: I fixed this by giving the `ins` element 100% width and a fixed height which fits the space, and then a pixel max width on the parent. This now works great :)

**Reply by checho221** [reply]
The method (window.adsbygoogle = window.adsbygoogle || []).push({}); is not working now, it is responding now with "All ins elements in the DOM with class=adsbygoogle already have ads in them" but it doesn't display any ad

**Reply by cateyelow** [reply]
@checho221 

this code works for me


```
  const router = useRouter();
  const adsLoaded = useRef(false);

  useEffect(() => {
    const loadAd = () => {
      if (typeof window !== "undefined" && window.adsbygoogle) {
        window.adsbygoogle = window.adsbygoogle || [];
        window.adsbygoogle.push({});
        adsLoaded.current = true;
      }
    };
  
    if (router.query && !adsLoaded.current) {
      setTimeout(loadAd, 0);
    }
  }, [router.query]);
```

**Reply by checho221** [reply]
It seems to be working better but not 100/100 router.query changes

**Reply by cateyelow** [reply]
@checho221 
normally, this code works well without router.query. it makes more dynamic ads when router changes.
if you use one ad-slot multiple times on one page, you should use 'id'
i gave you example, many ads that same slot on one page

[Ads.tsx]
```
import { useRouter } from 'next/router';
import { useEffect, useRef } from 'react';
import React from 'react';

declare global {
  interface Window {
    adsbygoogle: any;
  }
}

const GoogleAdPcItem = ({ adId }: { adId: string }) => {
  const router = useRouter();
  const adsLoaded = useRef<any>(false);

  useEffect(() => {
    const loadAd = () => {
      if (typeof window !== "undefined" && window.adsbygoogle) {
        window.adsbygoogle = window.adsbygoogle || [];
        window.adsbygoogle.push({});
        adsLoaded.current = true;
      }
    };
  
    if (router.query && !adsLoaded.current) {
      setTimeout(loadAd, 0);
    }
  }, [router.query]);

  return (
    <div 
      className="googleAd-container flex items-center justify-center border rounded-md"
      style={{ maxWidth: '282px', maxHeight: '282px' }}
    >
      <ins
        key={adId}
        ref={adsLoaded}
        id={adId}
        className="adsbygoogle h-full"
        style={{ display: 'inline-block', width: '282px', height: '282px'}}
        data-ad-client="ca-pub-1111111"
        data-full-width-responsive="true"
        data-ad-slot="11111111"
      />
    </div>
  );
};

export default GoogleAdPcItem;
```

[Page.tsx]
```
data.items.map((item: any, index: number) => {
 {index % 6 === 0 && index !== 0 && (
    <GoogleAdPcItem key={index+'ads'} adId={`ad-slot-${index}`} />
  )}
})
  ```

**Reply by GS-719** [reply]
> I dont have _document.js, I have index.js or _app.js?
Bro write <head>"YOUR SCRIPT TAG HERE"</head> in layout.js

**Comment by khuong291** [comment]


**Reply by thbestforyourbizdeployment** [reply]
Got the same issue

**Comment by kas** [comment]
~~I tried adding the AdSense code into my index's head section. But next build fails. Any ideas?~~

```
...
<Head>
  <script async src="https://pagead2.googlesyndication.com/pagead/js/adsbygoogle.js"></script>
  <script>
    (adsbygoogle = window.adsbygoogle || []).push({
      google_ad_client: "ca-pub-#",
      enable_page_level_ads: true
    });
  </script>
</Head>
...
```

```
./pages/index.js:15:28
Syntax error: Unexpected token, expected "}"

  13 |         <script>
  14 |           (adsbygoogle = window.adsbygoogle || []).push({
> 15 |             google_ad_client: "ca-pub-#",
     |                             ^
  16 |             enable_page_level_ads: true
  17 |           });
  18 |         </script>
```

~~I'm curious if anyone has gotten around this error, seeing as AdSense is asking me to add this code in the head section?~~

~~I've tried putting the code on a single line, I've tried wrapping the property names with quotes, but no luck.~~

**Sigh.** I refreshed the AdSense website and it gave me this code to paste instead:
```
<script data-ad-client="ca-pub-#" async src="https://pagead2.googlesyndication.com/pagead/js/adsbygoogle.js"></script>
```

**Reply by thbestforyourbizdeployment** [reply]
same issue

**Reply by shriekdj** [reply]
try this instead `(window.adsbygoogle = window.adsbygoogle || []).push({
      google_ad_client: "ca-pub-#",
      enable_page_level_ads: true
    });`

**Comment by sangolariel** [comment]
why ad does not show on my site? Do I need to wait for google to review my account?

**Reply by ThomasDLi** [reply]
Yes, google needs time to review your site and make sure it complies with their terms before they enable ads onto your site.

**Reply by sangolariel** [reply]
Do I need to have a full site? I deployed my site but only some page was the release? Some are still developing.

**Reply by ThomasDLi** [reply]
Google reviews whatever you have on the site at the time of you submitting the request.

**Reply by sangolariel** [reply]
okay, thank you so much.

**Comment by ghost** [comment]
hello there, the example explanation above not really works. I still get this error in the console:
`adsbygoogle.js?client=ca-pub-xxxxxxxxxxxxxxxxxxx AdSense head tag doesn't support data-nscript attribute.`
Any idea ?

**Reply by victoriaSh** [reply]
Loading AdSense as usual <script> in Head of _document.js, not 'next/script' helps. But I'm not sure if it's better.

**Reply by shriekdj** [reply]
that's is just and console warning. because you must be using Next's Script Component. NextJS add that attribute named data-nscript to our rendered `Script` component. It is neglegible. but if it causing any issues then try this props `strategy="lazyOnload"
          crossOrigin="anonymous"`.

**Comment by XahidEx** [comment]


**Reply by XahidEx** [reply]
Finally, Fixed. (Don't use any script components)

**Reply by ghost** [reply]
also on route change with nextjs ?

**Reply by XahidEx** [reply]
When you apply for Adsense, Don't use any script components. And if you get approved, Now you can use the "AfterInteractive" component.
I got my Adsense recently after facing many issues.

**Reply by ghost** [reply]
i did, but still sporadically black out or white screen if i change the route, would you kindly share a demo somewhere thanks

**Reply by XahidEx** [reply]
Ping me on Twitter. @xahidex

**Reply by datteroandrea** [reply]
Did removing scripts from the website got you approved from Goodle Adsense? Also could you please share your website just to know what I am doing wrong?

**Reply by shriekdj** [reply]
try `lazyOnload` strategy

**Comment by Mr-Afonso** [comment]
For people still struggling with this topic. Just follow this article: https://xahidex.com/blog/google-adsense
I get approved in 1 day. Thank you @XahidEx

**Reply by XahidEx** [reply]
Thanks a lot. Glad to hear that. 🙈💖

**Reply by codecret** [reply]
> For people still struggling with this topic. Just follow this article: https://xahidex.com/blog/google-adsense I get approved in 1 day. Thank you @XahidEx

hey can i contact you guys in instagram actually i did all this but wanna know should I add meta tags to reveal my website at google search results ?

**Reply by XahidEx** [reply]
Sure. My username is the same. (you can also contact me on Telegram)

**Reply by codecret** [reply]
> Sure. My username is the same. (you can also contact me on Telegram)

Done

**Reply by sahilbisht0602** [reply]
Hey for me ads are not coming is this might be an issue ? i am using 2 google adsense account....also by using this adsense approach we don't have to write below code in document.tsx ?

`useEffect(() => {
    var ads = document.getElementsByClassName("adsbygoogle").length;
    for (var i = 0; i < ads; i++) {
      try {
        (window.adsbygoogle = window.adsbygoogle || []).push({});
      } catch (e) {}
    }
  }, []);`

**Reply by XahidEx** [reply]
If you're using two AdSense on the same website, you should wait for your AdSense suspension. 😁
It's illegal. Maybe you already got the notice. Check your Adsense account.

**Reply by sahilbisht0602** [reply]
what !! earlier i have integrated ads on my web using third party but now i want to do by my own....so i have to remove there script first ?

**Reply by XahidEx** [reply]
Check my reply in another comment here. Follow the steps I've explained. And NEVER use two Adsense account ads on one website.

**Comment by XahidEx** [comment]
If anyone is facing any other issues please follow the article, [NextJs Adsense](https://xahidex.com/blog/nextjs-adsense)  mentioned above. **(Including NextJs13 App dir)**
If anyone else facing any issues other than these, please let me know.

**Comment by billelbacha4555555** [comment]
Hy everyone,
I Have The Same Problem I Follow Those Steps But Nothing ....

**Reply by XahidEx** [reply]
I think you didn't follow the method I mentioned in the previous comment.

**Reply by msdoliya** [reply]
what if i am not using typescript in next.js

**Reply by cateyelow** [reply]
@msdoliya 
just remove types on parameters :)
